skypalwarmup/JSUserRegistrationService/src/test/java/application/controller/UserControllerTest.java

```diff
@@ -0,0 +1,134 @@
+public class UserControllerTest {
+
+    @Autowired
+    private WebApplicationContext wac;
+
+    @Autowired
+    private UserService userService;
+
+    @Autowired
+    private MockUserServiceImpl mockUserService;
+
+    @TestConfiguration
+    static class MockUserServiceImplTestConfiguration {
+
+        @Bean
+        @Scope(ConfigurableBeanFactory.SCOPE_SINGLETON)
+        public UserService userService() {
+            return new MockUserServiceImpl();
+        }
+    }
+
+    private MockMvc mockMvc;
+
+    private User userUnderTest;
+
+    private ObjectMapper om = new ObjectMapper();
+
+    private static final String registrationActionUrlTemplate = "/api/users";
+
+    @Before
+    public void setUp() {
+        this.mockMvc = MockMvcBuilders.webAppContextSetup(this.wac).build();
+        mockUserService.deleteAll();
+        createUserUnderTest();
+        System.out.println(userService);
+    }
+
+    @After
+    public void tearDown() {
+        mockUserService.deleteAll();
+    }
+
+    // TEST HELPERS
+
+    private void createUserUnderTest() {
+        AccountInformation inTestAccountInformation = new AccountInformation();
+        inTestAccountInformation.setUserName(TestData.User.AccountInformation.USERNAME);
+        inTestAccountInformation.setPassword(TestData.User.AccountInformation.PASSWORD);
+
+        Contacts inTestContacts = new Contacts();
+        inTestContacts.setEmail(TestData.User.Contacts.EMAIL);
+        inTestContacts.setCellPhone(TestData.User.Contacts.PHONE);
+
+        PersonalDetails inTestPersonalDetails = new PersonalDetails();
+        inTestPersonalDetails.setAge(TestData.User.PersonalDetails.age);
+        inTestPersonalDetails.setFirstName(TestData.User.PersonalDetails.firstName);
+        inTestPersonalDetails.setLastName(TestData.User.PersonalDetails.lastName);
+
+        userUnderTest = new User();
+        userUnderTest.setAccountInformation(inTestAccountInformation);
+        userUnderTest.setContacts(inTestContacts);
+        userUnderTest.setPersonalDetails(inTestPersonalDetails);
+    }
+
+    private String getUserValueAsJSONString() throws Exception {
+        return om.writeValueAsString(userUnderTest);
+    }
+
+    // TEST CASES
+
+    @Test
+    public void register_whenUserDoesNotExist_returnsUserCreated() throws Exception {
+        this.mockMvc.perform(MockMvcRequestBuilders.post(registrationActionUrlTemplate).content(getUserValueAsJSONString())
+            .contentType(MediaType.APPLICATION_JSON))
+            .andExpect(MockMvcResultMatchers.status().isCreated())
+            .andExpect(MockMvcResultMatchers.content().string(RegistrationActionResponses.CREATED.statusMessage))
+            .andDo((result) -> Assert.assertTrue(mockUserService.exists(userUnderTest)))
+            .andReturn();
+    }
+
+    @Test
+    public void register_withInvalidEmail_returnsBadRequest() throws Exception {
+        userUnderTest.getContacts().setEmail("not an email");
```

Comment:
Please add the string "not an email" as a constant to TestConstants like the rest

Reply:
Done